wger/manager/views/schedule_step.py

```diff
@@ -61,12 +62,43 @@ def get_form_class(self):
         class StepForm(ModelForm):
             workout = ModelChoiceField(queryset=Workout.objects.filter(user=self.request.user))
 
+            def clean_duration(self):
+                """Ensures duration does not exceed period"""
+                duration = self.cleaned_data['duration']
+                schedule = Schedule.objects.get(pk=self.schedule_id)
+                sum_duration = schedule.schedulestep_set.all().aggregate(models.Sum('duration'))
+
+                # Catch NoneType error for 1st workout to be added to schedule.
+                sum_duration = sum_duration['duration__sum'] or 0
+                derived_duration = duration + sum_duration
+
+                # Check if duration exceeds cycle duration
+                if schedule.period == 'Macrocycle' and derived_duration > 52:
+                    raise ValidationError(_('Invalid duration- cycle duration exceeding 1 year'))
+                elif schedule.period == 'Mesocycle' and derived_duration > 6:
+                    raise ValidationError(_('Invalid duration- cycle duration exceeding 6 weeks'))
+                elif schedule.period == 'Microcycle' and derived_duration > 1:
+                    raise ValidationError(_('Invalid duration- cycle duration exceeding 1 week'))
+
+                return duration
+
```

Comment:
Great work on this function, well documented and adhering to coding conventions set by the team.